### PR TITLE
NFT transfer tracking

### PR DESCRIPTION
**Pull request by pro-wh**
stores an 'owner' field with the destination of the last transfer. it's anticipated that "burns" will transfer to 0x00...00 and leave it there. our record of the nft will still exist

### CODE REVIEW

File: analyzer/runtime/extract.go [hunk 1]
```diff
@@ -897,6 +917,7 @@ func extractEvents(blockData *BlockData, relatedAccountAddresses map[apiTypes.Ad
 						pt.Mutated = true
 					}
 					registerNFTExist(blockData.PossibleNFTs, eventAddr, tokenID)
+					registerNFTTransfer(blockData.PossibleNFTs, eventAddr, tokenID, toAddr)
```
Comment (ptrus) [inline analyzer/runtime/extract.go:920]: Afaik ERC-721 allows transferring nft to yourself (e.g. from address=to address). Is it ok that num-transfers is increased in that case as well?

**Review by ptrus — APPROVED** [review]

File: analyzer/runtime/extract.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/runtime/extract.go:920]: yeah sure. it doesn't seem like a high obstacle if for example people have to transfer it between two of their own accounts instead

File: analyzer/queries/queries.go [hunk 2]
```diff
@@ -601,10 +601,13 @@ var (
...
-      (runtime, token_address, nft_id, last_want_download_round)
+      (runtime, token_address, nft_id, owner, num_transfers, last_want_download_round)
     VALUES
-      ($1, $2, $3, $4)
-    ON CONFLICT (runtime, token_address, nft_id) DO NOTHING`
+      ($1, $2, $3, $4, $5, $6)
+    ON CONFLICT (runtime, token_address, nft_id) DO UPDATE
+    SET
+      owner = COALESCE(excluded.owner, chain.evm_nfts.owner),
+      num_transfers = chain.evm_nfts.num_transfers + excluded.num_transfers`
```
Comment (mitjat) [inline analyzer/queries/queries.go:610]: Optional: For upserts, I've recently been using an `old` alias like so, and liking it:
```sql
INSERT INTO chain.my_long_named_table AS old
...
ON CONFLICT DO UPDATE SET
  num_transfers = old.num_transfers + excluded.num_transfers
```

File: analyzer/queries/queries.go [hunk 3]
```diff
@@ -601,10 +601,13 @@ var (
 
 	RuntimeEVMNFTInsert = `
```
Comment (mitjat) [inline analyzer/queries/queries.go:602]:
```suggestion
	RuntimeEVMNFTUpsert = `
```

**Review by mitjat — APPROVED** [review]
:metal:

File: analyzer/queries/queries.go (hunk 2 above)
Comment (pro-wh) [reply analyzer/queries/queries.go:635]: k let me add that too

File: analyzer/runtime/extract.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/runtime/extract.go:990]: ok tentatively on the mint and burn not counting in num_transfers

File: analyzer/queries/queries.go (hunk 3 above)
Comment (pro-wh) [reply analyzer/queries/queries.go:627]: ):

Comment (pro-wh) [reply analyzer/queries/queries.go:627]: no on this. there's a lot of `...Insert` queries that define an ON CONFLICT. really we need better terminology for "the thing the block analyzer does" and "the thing the other analyzer does"

File: analyzer/runtime/extract.go (hunk 1 above)
Comment (pro-wh) [reply analyzer/runtime/extract.go:990]: ugh I just saw this other comment from the fungible tokens side

> Mints, burns, and zero-value transfers all count as transfers.

so I'm going to preserve consistency with that instead

Comment (pro-wh) [reply analyzer/runtime/extract.go:990]: and I'm opposed to fussing with the abstraction in the other suggestions about inlining registerNFTTransfer and eliminating registerNFTExist

File: analyzer/queries/queries.go (hunk 3 above)
Comment (pro-wh) [reply analyzer/queries/queries.go:627]: actually nah changing it is fine. most of the insert on conflict's are DO NOTHING

File: analyzer/runtime/extract.go (hunk 1 above)
Comment (mitjat) [reply analyzer/runtime/extract.go:990]: Thanks for finding the prior "art" on countin mints/burns. Might be worth mentioning in a comment in our codebase.

Tiny refactor/abstraction: sure, sg.

File: analyzer/queries/queries.go (hunk 3 above)
Comment (mitjat) [reply analyzer/queries/queries.go]: :+1: if they're not `DO NOTHING`, I'm in favor of renaming them too. Consistency is always nice. But regardless, I find the "upsert" in the name helpful when reading analyzer code; sometimes that clue is enough to guess/remember what the SQL query does, without jumping to the definition.

**Comment by pro-wh** [comment]
image

**Pull request merged**